tests/integration/Tribe/Attendee_Registration/Theme_Template_Test.php

```diff
@@ -0,0 +1,349 @@
+		$this->set_class_fn_return( WP_Query::class, 'is_main_query', $is_main_query );
+
+		// Mock `is_on_ar_page` and `is_on_custom_ar_page`.
+		$this->set_class_fn_return( Template::class, 'is_on_ar_page', $is_on_ar_page );
+		$this->set_class_fn_return( Template::class, 'is_on_custom_ar_page', $is_on_custom_ar_page );
+
+		// Create an instance of the class to test.
+		$template_instance = new Template();
+
+		// Run the method under test.
+		$result = $template_instance->setup_context( $posts, $query );
+
+		// Assert the result matches the expectation.
+		$this->assertEquals( $expected, $result, 'setup_context result should match the expected output.' );
+
+		// Delete the test post.
+		wp_delete_post( $posts[0]->ID, true );
+	}
+
+	/**
+	 * Data provider for `setup_context`.
+	 *
+	 * @return Generator
+	 */
+	public function setup_context_provider(): Generator {
+		yield 'Not the main query' => [
+			function () {
+				$post_id = self::factory()->post->create();
+				$post    = get_post( $post_id );
+
+				return [
+					[ $post ], // Posts.
+					new WP_Query(), // Query object.
+					false, // is_main_query.
+					false, // is_on_ar_page.
+					false, // is_on_custom_ar_page.
+					[ $post ], // Expected result (unchanged).
+				];
+			},
+		];
+
+		yield 'Main query, not on AR page' => [
+			function () {
+				$post_id = self::factory()->post->create();
+				$post    = get_post( $post_id );
+
+				return [
+					[ $post ], // Posts.
+					new WP_Query(), // Query object.
+					true, // is_main_query.
+					false, // is_on_ar_page.
+					false, // is_on_custom_ar_page.
+					[ $post ], // Expected result (unchanged).
+				];
+			},
+		];
+
+		yield 'Main query, not on AR page, multiple posts' => [
+			function () {
+				// Create 5 posts.
+				$posts = [];
+				for ( $i = 0; $i < 5; $i++ ) {
+					$post_id = self::factory()->post->create();
+					$posts[] = get_post( $post_id );
+				}
+
+				return [
+					$posts, // Posts array.
+					new WP_Query(), // Query object.
+					true, // is_main_query.
+					false, // is_on_ar_page.
+					false, // is_on_custom_ar_page.
+					$posts, // Expected result (unchanged).
+				];
+			},
+		];
+
+
+		yield 'On AR page, on custom AR page' => [
+			function () {
+				$post_id = self::factory()->post->create();
+				$post    = get_post( $post_id );
+
+				return [
+					[ $post ], // Posts.
+					new WP_Query(), // Query object.
+					true, // is_main_query.
+					true, // is_on_ar_page.
+					true, // is_on_custom_ar_page.
+					[ $post ], // Expected result (unchanged).
+				];
+			},
+		];
+
+		yield 'On AR page, not on custom AR page' => [
+			function () {
+				$template     = new Template();
+				$spoofed_page = $template->spoofed_page();
+
+				return [
+					[ $spoofed_page ], // Posts.
+					new WP_Query(), // Query object.
+					true, // is_main_query.
+					true, // is_on_ar_page.
+					false, // is_on_custom_ar_page.
+					[ $spoofed_page ], // Expected spoofed post.
```

Comment:
I would like to see giving it $posts and the method filters them out and returns the [ $spoofed_page ] instead as well